### PR TITLE
Added functionality of label coloring

### PR DESCRIPTION
**Pull request by konadrian**
LabelColor

### CODE REVIEW

File: app/Colors.php [hunk 1]
```diff
@@ -341,4 +345,32 @@ public static function getContrast($hexcolor)
...
 		return hexdec($hexcolor) > 0xffffff / $contrastRatio ? 'black' : 'white';
 	}
+
+	/**
+	 * Update field color code and generate stylesheet file.
+	 *
+	 * @param int    $ieldId
+	 * @param string $color
+	 */
+	public static function updateFieldColor($ieldId, $color)
```
Comment (rskrzypczak) [inline app/Colors.php]:
```suggestion
	public static function updateFieldColor($fieldId, $color)
```

File: app/Colors.php [hunk 2]
```diff
@@ -341,4 +345,32 @@ public static function getContrast($hexcolor)
...
+		static::generate('field');
+	}
+
+	/**
+	 * Generate fields colors stylesheet.
+	 */
+	private static function generateFields()
+	{
+		$css = '';
+		$query = (new \App\Db\Query())->select(['tabid', 'fieldname', 'color'])->from('vtiger_field');
```
Comment (rskrzypczak) [inline app/Colors.php]: only active

File: layouts/basic/modules/OSSPasswords/Detail/BlockView.tpl [hunk 3]
```diff
@@ -40,7 +40,7 @@
 				{/if}
 				<div class="col-sm-6">
 					<div class="form-row">
-						<div class="fieldLabel border-top border-left col-sm-6  {$WIDTHTYPE}"
+						<div class="labelColor_{$MODULE}_{$FIELD_MODEL->getFieldName()} fieldLabel border-top border-left col-sm-6  {$WIDTHTYPE}"
```
Comment (rskrzypczak) [inline layouts/basic/modules/OSSPasswords/Detail/BlockView.tpl]: labelColor? :)

File: layouts/basic/modules/OSSPasswords/EditViewBlocks.tpl [hunk 4]
```diff
@@ -94,7 +94,7 @@
 								{/if}
 								<div class="{if $FIELD_MODEL->getUIType() neq "300"} col-sm-6 form-row align-items-center my-1 mx-0 {else} w-100 {/if}">
 									{if $FIELD_MODEL->getUIType() neq "300"}
-										<div class="col-lg-12 col-xl-3 text-lg-left text-xl-right fieldLabel  {$WIDTHTYPE}">
+										<div class="labelColor_{$MODULE}_{$FIELD_MODEL->getFieldName()} col-lg-12 col-xl-3 text-lg-left text-xl-right fieldLabel  {$WIDTHTYPE}">
```
Comment (rskrzypczak) [inline layouts/basic/modules/OSSPasswords/EditViewBlocks.tpl]:
```suggestion
										<div class="labelColor_{$MODULE_NAME}_{$FIELD_MODEL->getFieldName()} col-lg-12 col-xl-3 text-lg-left text-xl-right fieldLabel  {$WIDTHTYPE}">
```

File: layouts/basic/modules/Settings/Colors/TabFieldColors.tpl [hunk 5]
```diff
@@ -1,68 +1,55 @@
...
-										<button data-fieldId="{$SELECTED_PICKLIST_FIELD_ID}" data-fieldValueId="{$PICKLIST_VALUE['picklistValueId']}" data-type="PicklistItem" class="btn btn-sm btn-primary mr-1 float-right updatePicklistValueColor"><span class="fas fa-edit"></span> {\App\Language::translate('LBL_UPDATE_COLOR',$QUALIFIED_MODULE)}</button>&ensp;
-										<button data-fieldId="{$SELECTED_PICKLIST_FIELD_ID}" data-fieldValueId="{$PICKLIST_VALUE['picklistValueId']}" data-type="PicklistItem" class="btn btn-sm btn-warning mr-1 float-right generatePicklistValueColor"><span class="fas fa-redo-alt"></span> {\App\Language::translate('LBL_GENERATE_COLOR',$QUALIFIED_MODULE)}</button>&ensp;
-									</td>
-								</tr>
-							{/foreach}
+						{foreach from=$SELECTED_MODULE_FIELDS item=$FIELD}
+							{assign var=FIELD_ID value=$FIELD->getId()}
+							<tr data-id="{$FIELD_ID}" data-color="{$FIELD->get('color')}">
+								<td>{\App\Language::translate($FIELD->getFieldLabel(),$SELECTED_MODULE_NAME)}</td>
+								<td class="js-color-preview{$FIELD_ID} calendarColor" data-color="{$FIELD->get('color')}" data-field-id="{$FIELD_ID}" style="background: {$FIELD->get('color')};"></td>
```
Comment (rskrzypczak) [inline layouts/basic/modules/Settings/Colors/TabFieldColors.tpl]: js-color-preview ?

File: app/Colors.php (hunk 1 above)
Comment (rskrzypczak) [inline app/Colors.php]:
```suggestion
	public static function updateFieldColor($ieldId, $color): bool
```

File: app/Colors.php [hunk 6]
```diff
@@ -341,4 +345,39 @@ public static function getContrast($hexcolor)
...
+	 */
+	public static function updateFieldColor($fieldId, $color): bool
+	{
+		$fieldHasColor = (bool) (new \App\Db\Query())->select(['color'])->from('vtiger_field')->where(['fieldid' => $fieldid])->scalar();
+		$updateFieldColorResult = (bool) Db::getInstance()->createCommand()->update('vtiger_field', ['color' => $color], ['fieldid' => $fieldId])->execute();
+		static::generate('field');
+		if ($updateFieldColorResult || !$fieldHasColor) {
+			return true;
+		}
+		return false;
```
Comment (rskrzypczak) [inline app/Colors.php]: one return
one query

File: modules/Settings/Colors/actions/SaveAjax.php [hunk 7]
```diff
@@ -283,4 +288,44 @@ public function removeCalendarColor(\App\Request $request)
...
+		$fieldId = $request->getInteger('fieldId');
+		if (!$request->has('color')) {
+			$color = \App\Colors::getRandomColor();
+		} else {
+			$color = $request->getByType('color', 'Color');
+		}
+		\App\Colors::updateFieldColor($fieldId, $color);
+		$response = new Vtiger_Response();
+		$response->setResult([
+			'success' => true,
```
Comment (rskrzypczak) [inline modules/Settings/Colors/actions/SaveAjax.php]: ?

File: modules/Settings/Colors/actions/SaveAjax.php [hunk 8]
```diff
@@ -283,4 +288,44 @@ public function removeCalendarColor(\App\Request $request)
...
+	 * Remove field color.
+	 *
+	 * @param \App\Request $request
+	 */
+	public function removeFieldColor(App\Request $request)
+	{
+		\App\Colors::updateFieldColor($request->getInteger('fieldId'), '');
+		$response = new Vtiger_Response();
+		$response->setResult([
+			'success' => true,
```
Comment (rskrzypczak) [inline modules/Settings/Colors/actions/SaveAjax.php]: ?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: app/Colors.php (hunk 6 above)
Comment (konadrian) [reply app/Colors.php]: That was made because I would to know if will update a record which is exists because I will get 0
